### ISSUE TITLE
DnsClient recursive resolution exceeded the maximum stack count

### ISSUE THREAD

**Issue by necrogami** [TechnitiumSoftware/DnsServer#271]
[2021-06-22 18:29:00 UTC] DNS Server recursive resolution failed for QNAME: blinkstick.com; QTYPE: A; QCLASS: IN;
TechnitiumLibrary.Net.Dns.DnsClientException: DnsClient recursive resolution exceeded the maximum stack count for domain: blinkstick.com
   at TechnitiumLibrary.Net.Dns.DnsClient.RecursiveResolveAsync(DnsQuestionRecord question, IDnsCache cache, NetProxy proxy, Boolean preferIPv6, Boolean randomizeName, Boolean qnameMinimization, Int32 retries, Int32 timeout, Int32 maxStackCount) in Z:\Technitium\Projects\TechnitiumLibrary\TechnitiumLibrary.Net\Dns\DnsClient.cs:line 368
   at DnsServerCore.Dns.DnsServer.RecursiveResolveAsync(DnsDatagram request, IReadOnlyList`1 viaForwarders, Boolean cachePrefetchOperation, Boolean cacheRefreshOperation, TaskCompletionSource`1 taskCompletionSource) in Z:\Technitium\Projects\DnsServer\DnsServerCore\Dns\DnsServer.cs:line 1993

Any  idea why i'm getting this? I'm using the root for recursion not through any public recursing server

**Comment by ShreyasZare** [comment]
Thanks for reporting this. I tried to resolve the domain at my end and sometimes its resolving well while sometimes giving the error you posted. So, I ran wireshark and tried it again. It seems that the domain name's DNS provider 1and1 has badly configured their name servers causing resolution loops. It depends on which name server is chosen randomly by the DNS server. For some selected name servers in the worst case scenario, resolving their IP address goes into loop which hits the limit on number of hops that can be taken.

The following is the screenshot when it gives the error. See how the name server resolution is heading from one to another to yet another back to the first.
![image](https://user-images.githubusercontent.com/10008172/123048420-03089900-d41c-11eb-9e3f-de82f152794b.png)

This is another screenshot where the DNS server managed to resolve the domain name finally just before hitting the max hop limit.

![image](https://user-images.githubusercontent.com/10008172/123048601-33503780-d41c-11eb-9956-4b02c3480ea2.png)

However, once the name server's IP address is resolved and cached, the domain name is resolving well. And when the NS records expire after 2 days, this error may come up again until the NS records are resolved again and cached.
